third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -15,152 +15,102 @@ namespace ttng = ::mlir::triton::nvidia_gpu;
-  // memory allocation.
-  // 3. ttg.MemDescIndexOp: An index into a memory allocation.
-  Operation *_allocOp = nullptr;
-
-public:
-  TMEM1DAllocator(OpBuilder &builder) : builder(builder) {}
-
-private:
-  void copyAttrs(Operation *oldOp, Operation *newOp) {
-    // Right now we just copy over ttg.partition per
-    // the example.
-    // TODO: Should we copy over loop information?
-    auto partitionAttr = oldOp->getAttr("ttg.partition");
-    newOp->setAttr("ttg.partition", partitionAttr);
-  }
-
-  void setExpandedInput(tt::ExpandDimsOp expandedInput) {
-    this->_expandedInput = expandedInput;
-  }
-
-  tt::ExpandDimsOp getExpandedInput() {
-    assert(_expandedInput != nullptr && "Must call setExpandedInput");
-    return _expandedInput;
-  }
-
-  void setAllocOp(Operation *allocOp) { this->_allocOp = allocOp; }
-
-  Operation *getAllocOp() {
-    assert(_allocOp != nullptr && "Must call getAllocOp()");
-    return _allocOp;
-  }
+//
+
+void TMEM1DAllocator::replaceWith1DTMEM(Operation *producer,
+                                        Operation *consumer,
+                                        Operation *allocOpBuffer) {
+  this->numWarps = ttg::lookupNumWarps(producer);
+  assert((numWarps == 4 || numWarps == 8) && "Only support 4 or 8 warps");
+  TMEMStore1D(producer, allocOpBuffer);
+  TMEMLoad1D(producer, consumer);
+}
 
-  RankedTensorType getResultTensorType(Operation *op, size_t expectedSize) {
-    auto outputType = dyn_cast<RankedTensorType>(op->getResult(0).getType());
-    if (!outputType || outputType.getShape().size() != 2) {
-      assert("Invalid tensor input");
-    }
-    return outputType;
-  }
+ttng::TMEMAllocOp TMEM1DAllocator::alloc1DTMEMBuffer() {
+  auto expandedInput = getExpandedInput();
+  auto oldRetType = getResultTensorType(expandedInput, 2);
+  auto shape = oldRetType.getShape();
+  auto tmemDesc = createTMEMDesc(builder, oldRetType, shape[0], shape[1]);
+  auto allocCall = builder.create<ttng::TMEMAllocOp>(
+      expandedInput->getLoc(), tmemDesc, builder.getType<ttg::AsyncTokenType>(),
+      /*src=*/Value());
+  return allocCall;
+}
 
-  ttng::TMEMAllocOp alloc1DTMEMBuffer() {
-    auto expandedInput = getExpandedInput();
-    auto oldRetType = getResultTensorType(expandedInput, 2);
-    auto shape = oldRetType.getShape();
-    auto tmemDesc = createTMEMDesc(builder, oldRetType, shape[0], shape[1]);
-    auto allocCall = builder.create<ttng::TMEMAllocOp>(
-        expandedInput->getLoc(), tmemDesc,
-        builder.getType<ttg::AsyncTokenType>(),
-        /*src=*/Value());
-    return allocCall;
+void TMEM1DAllocator::TMEMStore1D(Operation *producer,
+                                  Operation *allocOpBuffer) {
+  // Expand from 1D -> 2D
+  auto oldRetType = getResultTensorType(producer, 1);
+  builder.setInsertionPointAfter(producer);
+  // TODO(njriasan): This only works because
+  // producer->getResult(0) already has a ttg.slice attribute.
```

Comment:
This will be a merge conflict to resolve. It should be fairly easy, but let me know if you have issues